File: src/ds_agent/application/dtos/task_contract.py
"""DTOs for task contract boundaries."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

from ds_agent.domain.entities.assumption_log import AssumptionLog
from ds_agent.domain.entities.dataset_manifest import DatasetManifest
from ds_agent.domain.entities.delivery_pack import (
    DeliveryChannel,
    DeliveryPack,
    DeliveryPackStatus,
)
from ds_agent.domain.entities.goal_brief import GoalBrief
from ds_agent.domain.entities.metric_spec import MetricSpec
from ds_agent.domain.entities.review_verdict import (
    ActionHint,
    ConfidenceBand,
    Issue,
    LayerResult,
    ReviewVerdict,
    VerdictCategory,
    VerdictResult,
)
from ds_agent.domain.entities.task_contract import (
    AutonomyBoundary,
    Budget,
    DataSourceGrant,
    DefinitionOfDone,
    DeliverableSpec,
    TaskContract,
    TaskContractStatus,
)
from ds_agent.domain.entities.task_contract_bundle import TaskContractBundle
from ds_agent.domain.value_objects.audience_persona import AudiencePersona
from ds_agent.domain.value_objects.authority_mode import AuthorityMode


class GoalBriefDraftDTO(BaseModel):
    """Input payload for creating a GoalBrief."""

    business_question: str
    ds_problem_statement: str
    hypothesis: str | None = None
    comparison_baseline: str
    decision_to_make: str
    expected_effort: str


class TaskContractDraftDTO(BaseModel):
    """Input payload for creating a task contract draft."""

    session_id: str
    contract_type: str
    business_goal: str
    goal_brief: GoalBriefDraftDTO
    required_deliverables: list[DeliverableSpec]
    allowed_data_sources: list[DataSourceGrant] = Field(default_factory=list)
    forbidden_data_patterns: list[str] = Field(default_factory=list)
    budget: Budget = Field(default_factory=Budget)
    autonomy: AutonomyBoundary = Field(default_factory=AutonomyBoundary)
    decision_owner: str | None = None
    decision_deadline: datetime | None = None
    definition_of_done: DefinitionOfDone | None = None
    authority: AuthorityMode | None = None
    audience: AudiencePersona | None = None
    mission: str | None = None
    created_by: str = "agent"


class TaskContractUpdateDTO(BaseModel):
    """Patch + transition request for an existing contract."""

    task_id: str
    expected_version: int = Field(ge=1)
    patch: dict[str, Any] = Field(default_factory=dict)
    transition_to: TaskContractStatus | None = None
    reason: str | None = None


class AssumptionInputDTO(BaseModel):
    """Input payload for a single assumption entry."""

    task_id: str
    statement: str
    rationale: str
    risk_level: Literal["low", "medium", "high"]
    asked_user: bool = False


class VerifyAssumptionDTO(BaseModel):
    """Input payload for marking one assumption as verified."""

    task_id: str
    entry_id: str
    expected_version: int = Field(ge=1)
    verification_note: str | None = None


class ReviewVerdictInputDTO(BaseModel):
    """Input payload for recording a review verdict."""

    task_id: str
    verdict_id: str | None = None
    category: VerdictCategory = "orchestrator"
    result: VerdictResult | None = None
    reviewer: str = "verifier"
    summary: str | None = None
    evidence_refs: list[str] = Field(default_factory=list)
    run_id: str | None = None
    layers: list[LayerResult] = Field(default_factory=list)
    blocking_issues: list[Issue] = Field(default_factory=list)
    confidence: ConfidenceBand | None = None
    recommended_actions: list[ActionHint] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)


class DeliveryPackInputDTO(BaseModel):
    """Input payload for recording a delivery pack."""

    task_id: str
    items: list[dict[str, Any]] = Field(default_factory=list)
    artifacts: list[dict[str, Any]] = Field(default_factory=list)
    follow_up_actions: list[str] = Field(default_factory=list)
    source_analysis_id: str | None = None
    confidence: float | None = Field(default=None, ge=0.0, le=1.0)
    signed_by: str | None = None
    signature: str | None = None
    global_context: dict[str, str] = Field(default_factory=dict)
    status: DeliveryPackStatus = DeliveryPackStatus.RENDERED
    tenant: str = Field(default="default", min_length=1)


class BuildDeliveryPackDTO(BaseModel):
    """Input payload for planning a typed delivery pack from a task contract."""

    task_id: str
    audiences: list[str] = Field(default_factory=list)
    follow_up_actions: list[str] = Field(default_factory=list)
    source_analysis_id: str | None = None
    confidence: float | None = Field(default=None, ge=0.0, le=1.0)
    signed_by: str | None = None
    signature: str | None = None
    global_context: dict[str, str] = Field(default_factory=dict)
    tenant: str = Field(default="default", min_length=1)


class RenderDeliveryArtifactDTO(BaseModel):
    """Input payload for rendering one artifact from a persisted delivery pack."""

    task_id: str
    artifact_id: str
    analysis: dict[str, Any] | str
    output_dir: str
    audience_profile: AudiencePersona | None = None


class DispatchDeliveryDTO(BaseModel):
    """Input payload for dispatching rendered delivery artifacts."""

    task_id: str
    artifact_ids: list[str] = Field(default_factory=list)
    channels: list[DeliveryChannel] = Field(default_factory=list)
    dry_run: bool = False
    approve_manual_review: bool = False


class ListDeliveryLogDTO(BaseModel):
    """Input payload for querying persisted delivery-dispatch logs."""

    task_id: str
    pack_id: str | None = None
    artifact_ids: list[str] = Field(default_factory=list)
    channels: list[DeliveryChannel] = Field(default_factory=list)
    limit: int = Field(default=50, ge=1, le=500)


class TaskContractListItemDTO(BaseModel):
    """Compact list view of a task contract."""

    task_id: str
    status: str
    type: str
    business_goal: str
    updated_at: datetime
    version: int

    @classmethod
    def from_contract(cls, contract: TaskContract) -> TaskContractListItemDTO:
        return cls(
            task_id=contract.task_id,
            status=contract.status.value,
            type=contract.type,
            business_goal=contract.business_goal,
            updated_at=contract.updated_at,
            version=contract.version,
        )


class TaskContractViewDTO(BaseModel):
    """Detailed view returned to tools and presentation layers."""

    contract: TaskContract
    goal_brief: GoalBrief | None = None
    metric_specs: list[MetricSpec] = Field(default_factory=list)
    dataset_manifest: DatasetManifest | None = None
    assumption_log: AssumptionLog | None = None
    review_verdicts: list[ReviewVerdict] = Field(default_factory=list)
    delivery_pack: DeliveryPack | None = None
    dod_summary: list[str] = Field(default_factory=list)

    @classmethod
    def from_bundle(
        cls,
        bundle: TaskContractBundle,
        *,
        include: set[str] | None = None,
        dod_summary: list[str] | None = None,
    ) -> TaskContractViewDTO:
        include_all = include is None
        include_set = include or set()
        return cls(
            contract=bundle.contract,
            goal_brief=(bundle.goal_brief if include_all or "goal_brief" in include_set else None),
            metric_specs=(
                bundle.metric_specs if include_all or "metric_specs" in include_set else []
            ),
            dataset_manifest=(
                bundle.dataset_manifest
                if include_all or "dataset_manifest" in include_set
                else None
            ),
            assumption_log=(
                bundle.assumption_log if include_all or "assumption_log" in include_set else None
            ),
            review_verdicts=(
                bundle.review_verdicts if include_all or "review_verdicts" in include_set else []
            ),
            delivery_pack=(
                bundle.delivery_pack if include_all or "delivery_pack" in include_set else None
            ),
            dod_summary=dod_summary or [],
        )

File: electron/src/renderer/types/taskContract.ts
export type TaskContractStatus =
  | 'draft'
  | 'agreed'
  | 'in_progress'
  | 'review'
  | 'closed'
  | 'abandoned';

export type DeliveryPackStatus = 'draft' | 'rendered' | 'dispatched' | 'rejected';

export type DeliveryDispatchMode = 'auto' | 'auto_with_signature' | 'manual_review';

export type DeliveryDispatchStatus =
  | 'sent'
  | 'blocked'
  | 'duplicate'
  | 'failed'
  | 'dry_run';

export type DeliveryDispatchOutcome = 'dispatched' | 'partial' | 'blocked' | 'dry_run';

export interface TaskContractListItem {
  task_id: string;
  status: TaskContractStatus;
  type: string;
  business_goal: string;
  updated_at: string;
  version: number;
}

export interface DeliverableSpec {
  type: string;
  audience: string;
  format: string;
  count?: number | null;
}

export interface GoalBriefView {
  business_question: string;
  ds_problem_statement: string;
  comparison_baseline: string;
  decision_to_make: string;
  hypothesis?: string | null;
  expected_effort: string;
}

export interface GoalBriefDraftInput {
  business_question: string;
  ds_problem_statement: string;
  comparison_baseline: string;
  decision_to_make: string;
  hypothesis?: string | null;
  expected_effort: string;
}

export interface TaskContractCreatePayload {
  session_id: string;
  contract_type: string;
  business_goal: string;
  goal_brief: GoalBriefDraftInput;
  required_deliverables: DeliverableSpec[];
  allowed_data_sources?: Record<string, unknown>[];
  forbidden_data_patterns?: string[];
  budget?: Record<string, unknown>;
  autonomy?: Record<string, unknown>;
  decision_owner?: string | null;
  decision_deadline?: string | null;
  definition_of_done?: Record<string, unknown> | null;
  authority?: string | null;
  audience?: string | null;
  mission?: string | null;
  created_by?: string;
}

export interface TaskContractCreateResultView {
  task_id: string;
  status: TaskContractStatus;
  goal_brief_id: string;
  new_version: number;
  next_suggested_action?: string | null;
}

export interface TaskContractErrorDetailView {
  message: string;
  error_code?: string | null;
  metadata?: Record<string, unknown>;
}

export interface TaskContractIpcErrorResult {
  ok: false;
  error: string;
  errorDetail?: TaskContractErrorDetailView;
}

export interface AssumptionEntry {
  entry_id: string;
  statement: string;
  rationale: string;
  risk_level: 'low' | 'medium' | 'high';
  verified: boolean;
  verification_note?: string | null;
  asked_user: boolean;
  created_at: string;
}

export interface AssumptionLogView {
  log_id: string;
  task_id: string;
  entries: AssumptionEntry[];
}

export interface AssumptionVerificationResultView {
  entry_id: string;
  verified: boolean;
  verification_note?: string | null;
  new_version: number;
}

export interface ReviewVerdictView {
  verdict_id: string;
  task_id: string;
  category: string;
  result: 'pass' | 'warn' | 'fail';
  reviewer: string;
  summary: string;
  evidence_refs: string[];
  created_at: string;
  confidence?: {
    score: number;
    grade: 'high' | 'medium' | 'low' | 'insufficient' | null;
    rationale: string;
  } | null;
  blocking_issues?: {
    layer?: string | null;
    check_id?: string | null;
    severity: 'low' | 'medium' | 'high' | 'critical';
    message: string;
    blocking: boolean;
  }[];
  layers?: {
    layer: string;
    overall?: 'pass' | 'warn' | 'fail' | 'error' | null;
    score?: number | null;
    summary?: string | null;
    metadata?: Record<string, unknown>;
  }[];
  metadata?: Record<string, unknown>;
}

export interface DeliveryItemView {
  deliverable_type: string;
  audience: string;
  format: string;
  artifact_path: string;
  checksum?: string | null;
  delivered: boolean;
  delivery_channel?: string | null;
  artifact_id?: string | null;
  verifier_report_id?: string | null;
}

export interface DeliveryReceiverView {
  role: string;
  resolver: 'org_directory' | 'static_list' | 'task_contract';
  static_addresses: string[];
}

export interface DeliveryContentPolicyView {
  structure: string[];
  max_pages?: number | null;
  chart_count_range?: [number, number] | null;
  technical_detail: 'minimal' | 'balanced' | 'deep';
  tone: 'decisive' | 'actionable' | 'precise' | 'neutral' | 'mentoring';
  include_code: boolean;
  include_verifier_results: boolean;
  include_jira_links: boolean;
  include_feature_registry_refs: boolean;
  speculative_claims: 'allowed' | 'flagged' | 'forbidden';
}

export interface DeliveryArtifactView {
  artifact_id: string;
  type: string;
  audience: string;
  format: string;
  content_policy: DeliveryContentPolicyView;
  template_ref: string;
  delivery_channel: string[];
  dispatch_mode: DeliveryDispatchMode;
  receivers: DeliveryReceiverView[];
  rendered_uri?: string | null;
  verifier_report_id?: string | null;
}

export interface DeliveryPackView {
  pack_id: string;
  task_id: string;
  source_analysis_id?: string | null;
  confidence?: number | null;
  signed_by?: string | null;
  signature?: string | null;
  global_context: Record<string, string>;
  artifacts: DeliveryArtifactView[];
  items: DeliveryItemView[];
  follow_up_actions: string[];
  generated_at: string;
  status: DeliveryPackStatus;
  tenant: string;
}

export interface DeliveryReceiptView {
  artifact_id: string;
  channel: string;
  status: DeliveryDispatchStatus;
  idempotency_key: string;
  recorded_at: string;
  reason?: string | null;
  receipt_id?: string | null;
  adapter_name?: string | null;
}

export interface DeliveryBuildResultView {
  pack_id: string;
  items: number;
  artifacts: number;
  audiences: string[];
  artifact_ids: string[];
  status: DeliveryPackStatus;
  new_version: number;
}

export interface DeliveryRenderResultView {
  pack_id: string;
  artifact_id: string;
  output_path: string;
  format: string;
  verifier_status: string;
  verifier_report_id?: string | null;
  flagged_claims: string[];
  pack_status: DeliveryPackStatus;
  new_version: number;
  renderer_mode?: string | null;
  renderer_model?: string | null;
}

export interface DeliveryDispatchResultView {
  task_id: string;
  pack_id: string;
  dispatch_status: DeliveryDispatchOutcome;
  dry_run: boolean;
  receipts: DeliveryReceiptView[];
  log_path?: string | null;
  sent: number;
  blocked: number;
  failed: number;
  duplicates: number;
  pack_status: DeliveryPackStatus;
  new_version: number;
}

export interface DeliveryLogRecordView {
  task_id: string;
  pack_id: string;
  artifact_id: string;
  channel: string;
  status: DeliveryDispatchStatus;
  idempotency_key: string;
  recorded_at: string;
  reason?: string | null;
  receipt_id?: string | null;
  adapter_name?: string | null;
}

export interface DeliveryLogSummaryView {
  task_id: string;
  pack_id: string;
  pack_status: string;
  artifact_count: number;
  rendered_count: number;
  sent: number;
  blocked: number;
  duplicate: number;
  failed: number;
  dry_run: number;
  last_attempt?: string | null;
}

export interface DeliveryLogQueryResultView {
  task_id: string;
  pack_id?: string | null;
  log_path?: string | null;
  records: DeliveryLogRecordView[];
  summary?: DeliveryLogSummaryView | null;
  returned: number;
}

export interface PptxPreviewSlideView {
  index: number;
  section?: string | null;
  title: string;
  bullets: string[];
  excerpt: string;
  chart_count: number;
}

export type RenderedArtifactPreviewView =
  | {
      kind: 'markdown';
      path: string;
      fileUrl: string;
      content: string;
      truncated: boolean;
      sections: string[];
      lineCount: number;
    }
  | {
      kind: 'ipynb';
      path: string;
      fileUrl: string;
      markdownCells: string[];
      codeCells: string[];
      outputCells: string[];
      truncated: boolean;
      cellCount: number;
    }
  | {
      kind: 'pdf';
      path: string;
      fileUrl: string;
      size: number;
    }
  | {
      kind: 'pptx';
      path: string;
      fileUrl: string;
      slides: PptxPreviewSlideView[];
      slideCount: number;
      manifestFound: boolean;
    }
  | {
      kind: 'unavailable';
      path: string;
      fileUrl: string;
      message: string;
    };

export type ShadowLegacyState = 'triggered' | 'clear' | 'not_applicable';

export type ShadowMismatchKind =
  | 'legacy_only'
  | 'verifier_only'
  | 'agreement'
  | 'not_applicable';

export interface ShadowComparisonItemView {
  comparison_key: string;
  legacy_source: string;
  verifier_targets: string[];
  applicable: boolean;
  legacy_state: ShadowLegacyState;
  verifier_state: ShadowLegacyState;
  matches: boolean;
  mismatch_kind: ShadowMismatchKind;
  legacy_evidence: Record<string, unknown>[];
  verifier_evidence: Record<string, unknown>[];
  note?: string | null;
}

export interface ShadowComparisonView {
  comparison_id: string;
  verdict_id: string;
  task_id: string;
  run_id?: string | null;
  session_id?: string | null;
  created_at: string;
  items: ShadowComparisonItemView[];
  applicable_count: number;
  mismatch_count: number;
  match_rate: number;
  metadata?: Record<string, unknown>;
}

export interface TaskContractEntity {
  task_id: string;
  session_id: string;
  type: string;
  status: TaskContractStatus;
  business_goal: string;
  decision_owner?: string | null;
  authority?: string | null;
  audience?: string | null;
  mission?: string | null;
  forbidden_data_patterns: string[];
  required_deliverables: DeliverableSpec[];
  autonomy: {
    agent_will_do: string[];
    agent_will_ask: string[];
    agent_will_escalate: string[];
  };
  created_at: string;
  updated_at: string;
  version: number;
}

export interface TaskContractView {
  contract: TaskContractEntity;
  goal_brief?: GoalBriefView | null;
  assumption_log?: AssumptionLogView | null;
  review_verdicts: ReviewVerdictView[];
  delivery_pack?: DeliveryPackView | null;
  dod_summary: string[];
}
